[BUG] cudf.read_parquet fails to read from remote storage with use_python_file_object=True

**Describe the bug**
As [discussed](https://github.com/rapidsai/cudf/pull/9589#issuecomment-960927214) in #9589, cudf fails to read from remote storage when there is a RangeIndex specified in the pandas metadata, and `use_python_file_object=True`. This is because cudf [will try to use pyarrow](https://github.com/rapidsai/cudf/blob/f041a47954cd3dd7498589dd190b0eac6b0fa169/python/cudf/cudf/_lib/parquet.pyx#L206) to process the pandas metadata after the file source has already been converted to `cudf._lib.io.datasource.NativeFileDatasource` (which pyarrow does not support).

**Steps/Code to reproduce bug**
The easiest way to reproduce the bug is to add `use_python_file_object=True` to the `read_parquet` call in [test_read_parquet_filters](https://github.com/rapidsai/cudf/blob/f041a47954cd3dd7498589dd190b0eac6b0fa169/python/cudf/cudf/tests/test_s3.py#L275). This will result in the following failure:

```
____________________________________________________________________ test_read_parquet_filters ____________________________________________________________________

s3_base = 'http://127.0.0.1:5000/', s3so = {'client_kwargs': {'endpoint_url': 'http://127.0.0.1:5000/'}}
pdf =    Integer  Float  Integer2 String  Boolean
0     2345  9.001      2345  Alpha     True
1    11987  8.343       106   Beta    False
2     9027  6.000      2088  Gamma     True
3     9027  2.781    789277  Delta    False

    def test_read_parquet_filters(s3_base, s3so, pdf):
        fname = "test_parquet_reader_filters.parquet"
        bname = "parquet"
        buffer = BytesIO()
        pdf.to_parquet(path=buffer)
        buffer.seek(0)
        filters = [("String", "==", "Omega")]
        with s3_context(s3_base=s3_base, bucket=bname, files={fname: buffer}):
>           got = cudf.read_parquet(
                "s3://{}/{}".format(bname, fname),
                storage_options=s3so,
                filters=filters,
                use_python_file_object=True,
            )

cudf/tests/test_s3.py:283: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/rzamora/miniconda/miniconda3/envs/cudf_21.10b/lib/python3.8/site-packages/cudf/io/parquet.py:396: in read_parquet
    return libparquet.read_parquet(
cudf/_lib/parquet.pyx:110: in cudf._lib.parquet.read_parquet
    cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
cudf/_lib/parquet.pyx:206: in cudf._lib.parquet.read_parquet
    pa.parquet.read_metadata(s) for s in filepaths_or_buffers
/home/rzamora/miniconda/miniconda3/envs/cudf_21.10b/lib/python3.8/site-packages/pyarrow/parquet.py:2232: in read_metadata
    return ParquetFile(where, memory_map=memory_map).metadata
/home/rzamora/miniconda/miniconda3/envs/cudf_21.10b/lib/python3.8/site-packages/pyarrow/parquet.py:228: in __init__
    self.reader.open(
pyarrow/_parquet.pyx:966: in pyarrow._parquet.ParquetReader.open
    ???
pyarrow/io.pxi:1531: in pyarrow.lib.get_reader
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   TypeError: Cannot convert cudf._lib.io.datasource.NativeFileDatasource to pyarrow.lib.NativeFile

pyarrow/io.pxi:1524: TypeError
===================================================================== short test summary info =====================================================================
FAILED cudf/tests/test_s3.py::test_read_parquet_filters - TypeError: Cannot convert cudf._lib.io.datasource.NativeFileDatasource to pyarrow.lib.NativeFile
```

**Expected behavior**
Reading a valid parquet file from remote storage should never fail with `use_python_file_object=True`

--- a/python/cudf/cudf/_lib/parquet.pyx
+++ b/python/cudf/cudf/_lib/parquet.pyx
@@ -118,9 +118,17 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
     cudf.io.parquet.read_parquet
     cudf.io.parquet.to_parquet
     """
+
+    # Convert NativeFile buffers to NativeFileDatasource,
+    # but save original buffers in case we need to use
+    # pyarrow for metadata processing
+    # (See: https://github.com/rapidsai/cudf/issues/9599)
+    pa_buffers = []
     for i, datasource in enumerate(filepaths_or_buffers):
         if isinstance(datasource, NativeFile):
+            pa_buffers.append(datasource)
             filepaths_or_buffers[i] = NativeFileDatasource(datasource)
+
     cdef cudf_io_types.source_info source = make_source_info(
         filepaths_or_buffers)
 
@@ -203,7 +211,9 @@ cpdef read_parquet(filepaths_or_buffers, columns=None, row_groups=None,
             range_index_meta = index_col[0]
             if row_groups is not None:
                 per_file_metadata = [
-                    pa.parquet.read_metadata(s) for s in filepaths_or_buffers
+                    pa.parquet.read_metadata(s) for s in (
+                        pa_buffers or filepaths_or_buffers
+                    )
                 ]
 
                 filtered_idx = []

--- a/python/cudf/cudf/tests/test_s3.py
+++ b/python/cudf/cudf/tests/test_s3.py
@@ -272,7 +272,8 @@ def test_read_parquet_arrow_nativefile(s3_base, s3so, pdf, columns):
     assert_eq(expect, got)
 
 
-def test_read_parquet_filters(s3_base, s3so, pdf):
+@pytest.mark.parametrize("python_file", [True, False])
+def test_read_parquet_filters(s3_base, s3so, pdf, python_file):
     fname = "test_parquet_reader_filters.parquet"
     bname = "parquet"
     buffer = BytesIO()
@@ -284,6 +285,7 @@ def test_read_parquet_filters(s3_base, s3so, pdf):
             "s3://{}/{}".format(bname, fname),
             storage_options=s3so,
             filters=filters,
+            use_python_file_object=python_file,
         )
 
     # All row-groups should be filtered out

